lib/services/privileges_service.ts

```diff
@@ -237,14 +237,63 @@ export class SNPrivilegesService extends PureService {
     };
   }
 
+  async authenticate(action: ProtectedAction): Promise<boolean> {
+    const requiredCredentials = await this.netCredentialsForAction(action);
+
+    if (requiredCredentials.length === 0) {
+      return true;
+    }
+
+    const challengePrompts = []
+
+    for (const credential of requiredCredentials) {
+      if (credential === PrivilegeCredential.AccountPassword) {
+        challengePrompts.push(new ChallengePrompt(
+          ChallengeValidation.AccountPassword,
+          undefined,
+          ChallengeStrings.AccountPasswordPlaceholder
+        ));
+      }
+      if (credential === PrivilegeCredential.LocalAuthentication) {
+        if (await this.hasPasscode()) {
+          challengePrompts.push(new ChallengePrompt(
+            ChallengeValidation.LocalPasscode,
+            undefined,
+            ChallengeStrings.LocalPasscodePlaceholder
+          ));
+        }
+        if (await this.hasBiometrics()) {
+          challengePrompts.push(new ChallengePrompt(
+            ChallengeValidation.Biometric
+          ));
+        }
+      }
+    }
+
+    challengePrompts.push(new ChallengePrompt(ChallengeValidation.Form, PromptTitles.SessionLength, undefined, false, undefined, this.getSessionLengthOptions()));
+
+    const response = await this.challengeService?.promptForChallengeResponse(new Challenge(challengePrompts, ChallengeReason.AuthenticatePrivilege, true));
+    if (!response) {
+      return false;
+    }
+    const value = response.getValueForType(ChallengeValidation.Form);
+    if (value) {
+      this.setSessionLength(value.value as number)
+    }
+    return true;
+  }
+
+   /**
+   * @deprecated
+   */
```

Comment:
Should probably include a comment about what should be used instead here as well